esp-wifi/src/compat/common.rs

```diff
@@ -47,46 +47,78 @@ impl RawQueue {
+                unsafe {
+                    let p = self.storage.byte_add(self.item_size * self.current_write);
+                    p.copy_from(item as *mut u8, self.item_size);
+                    self.current_write = (self.current_write + 1) % self.count;
+                }
+
+                1
+            } else {
+                0
             }
+        })
+    }
 
-            1
-        } else {
-            0
-        }
+    pub(crate) fn try_dequeue(&mut self, item: *mut c_void) -> bool {
+        critical_section::with(|_| {
+            if self.count() > 0 {
+                unsafe {
+                    let p =
+                        self.storage.byte_add(self.item_size * self.current_read) as *const c_void;
+                    item.copy_from(p, self.item_size);
+                    self.current_read = (self.current_read + 1) % self.count;
+                }
+                true
+            } else {
+                false
+            }
+        })
     }
 
-    fn try_dequeue(&mut self, item: *mut c_void) -> bool {
-        if self.count() > 0 {
-            unsafe {
-                let p = self.storage.byte_add(self.item_size * self.current_read) as *const c_void;
-                item.copy_from(p, self.item_size);
-                self.current_read = (self.current_read + 1) % self.count;
+    pub(crate) fn remove(&mut self, item: *mut c_void) {
+        // do what the ESP-IDF implementations does ...
+        // just remove all elements and add them back except the one we need to remove -
+        // good enough for now
+        critical_section::with(|_| unsafe {
+            let item_slice = core::slice::from_raw_parts_mut(item as *mut u8, self.item_size);
+            let count = self.count();
+            let tmp_item = crate::compat::malloc::malloc(self.item_size);
+
+            if tmp_item.is_null() {
+                panic!("Out of memory");
             }
-            true
-        } else {
-            false
-        }
+
+            for _ in 0..count {
+                if self.try_dequeue(tmp_item as *mut c_void) {
+                    let tmp_slice = core::slice::from_raw_parts_mut(tmp_item, self.item_size);
+                    if tmp_slice != item_slice {
+                        self.enqueue(tmp_item as *mut c_void);
+                    }
+                }
+            }
+            crate::compat::malloc::free(tmp_item);
+        })
     }
 
-    fn count(&self) -> usize {
-        if self.current_write >= self.current_read {
-            self.current_write - self.current_read
-        } else {
-            self.count - self.current_read + self.current_write
-        }
+    pub(crate) fn count(&self) -> usize {
+        critical_section::with(|_| {
+            if self.current_write >= self.current_read {
+                self.current_write - self.current_read
+            } else {
+                self.count - self.current_read + self.current_write
+            }
```

Comment:
Why isn't this simply 

```suggestion
            self.count
```

? And what do the `self.count() < self.count` conditions mean elsewhere? I feel like this needs some love :)

Reply:
hope the renaming makes this more clear

Reply:
Yes, it does, thank you